# DeepEAG: A deep learning-based hybrid framework for identifying epilepsy-associated genes using a stacking strategy
===========================================================================


[![license](https://img.shields.io/badge/python_-3.8.0_-blue)](https://www.python.org/)
[![license](https://img.shields.io/badge/torch_-1.12.0_-blue)](https://pytorch.org/)
[![license](https://img.shields.io/badge/scipy_-1.5.0_-blue)](https://scipy.org/)
[![license](https://img.shields.io/badge/xgboost_-1.0.2_-blue)](https://xgboost.readthedocs.io/en/stable/)
[![license](https://img.shields.io/badge/scikit_learn_-1.2.0_-blue)](https://scikit-learn.org/)
[![license](https://img.shields.io/badge/numpy_-1.23.5_-blue)](https://numpy.org/)

As one of the most common neurological diseases in the world, epilepsy can be caused through the deletion or duplication of known epilepsy-associated genes. Therefore, it is crucial to accurately identify epilepsy-associated genes to gain a deeper understanding of the pathogenesis of epilepsy and to develop new therapies. Although a number of wet-lab experimental methods have been proposed, they are reliable but in general time-consuming and labor-intensive. Hence, their practical application in epilepsy-associated genes identification is quite limited. To address the existing limitations, we propose a computational method, called DeepEAG, for the identification of epilepsy-associated genes throughout the human genome. To develop DeepEAG, we first constructed a new and, to our knowledge, the most comprehensive epilepsy genomic benchmark dataset. Afterwards, we investigated four feature encoding algorithms with different perspectives and trained them using well-established traditional classifiers and deep learning classifier, resulting in 13 baseline models. Finally, the stacking strategy is effectively utilized by integrating the predicted output of the optimal baseline models and training with xgboost. In results, the DeepEAG predictor obtained excellent performance on the cross-validation with accuracy and AUC of 0.835 and 0.907, respectively. Overall, DeepEAG showed more stable and accurate predictive performance compared to baseline models, and also outperforms other ensemble strategies, demonstrating the effectiveness of our proposed hybrid framework. In addition, DeepEAG is anticipated to facilitate community-wide efforts to identify putative epilepsy-associated genes and generate new biological insights and testable hypotheses. The schematic diagram of DeepEAG is shown as follows:


![Image text](https://github.com/JfXie/DeepEAG/blob/main/IMG/framework.png)

**(a)** Dataset construction. First, we collected epilepsy-associated genes from recent databases and papers, and obtained their curated sequence from Uniprot database. Second, the CD-HIT tool with 90\% similarity threshold was utilized for redundancy reduction. Third, we checked and removed incorrect amino acid representation for each sequence. Finally, a positive benchmark dataset about epilepsy-associated genes was obtained, which was the largest and most complete one up to date to the best of our knowledge. **(b)** Feature extraction. We investigated four feature encoding algorithms with different perspectives, including sequence-based feature encoding algorithm, physicochemical property-based feature encoding algorithm, evolutionary information-based feature encoding algorithm and one-hot feature encoding algorithm. **(c)** Model training and stacking construction. First, we employed four classifiers and respectively constructed their models based on the different features to predict the score (S1...S10) of the test sample being the epilepsy-associated gene, including deep learning neural network (DNN), adaboost, random forest (RF), and support-vector machine (SVM). Second, the prediction labels generated by the baseline models were concatenated and used as inputs for training the ultimate meta model employing the Xgboost algorithm. **(d)** Performance evaluation. We evaluated the performance of DeepEAG using 5-fold cross-validation method, and visualized the results by the Receiver Operating Characteristic and Precision-Recall curve.

# Table of Contents

- [Installation](#installation)
- [Data description](#data-description)
- [Quick start](#quick-start)
- [Contributing](#contributing)
- [Contacts](#contacts)
- [License](#license)


# Installation

DeepEAG is tested to work under:

```
* Python 3.8.0
* Torch 1.12.0
* Scipy 1.5.0
* Xgboost 1.0.2
* Scikit-learn 1.2.0
* Numpy 1.23.5
* Other basic python toolkits
```
# Data description

| File name  | Description |
| ------------- | ------------- |
| Clinvar_epilepsy_genes.fasta  | Epilepsy-associated genes obtained from ClinVar database and their sequences obtained from Uniprot database|
| DisGeNET_epilepsy_genes.fasta  | Epilepsy-associated genes obtained from DisGeNET database and their sequences obtained from Uniprot database  |
| EpilepsyGeneDatabase_epilepsy_genes.fasta  | Epilepsy-associated genes obtained from EpilepsyGeneDatabase database and their sequences obtained from Uniprot database   |
| Genes4Epilepsy_genes.fasta| Epilepsy-associated genes obtained from Genes4Epilepsy database and their sequences obtained from Uniprot database |
| GWAS_epilepsy_genes.fasta| Epilepsy-associated genes obtained from GWAS database and their sequences obtained from Uniprot database| 
| Uniprot_epilepsy_genes.fasta|  Epilepsy-associated genes and their sequences obtained from Uniprot database| 
| Wang_epilepsy_genes.fasta| Epilepsy-associated genes obtained from Wang et al. and their sequences obtained from Uniprot database| 


# Quick start
To reproduce our results:

## 1，Data preprocessing and construction of multi-source heterogeneous biomolecular networks
```
python ./src/data_process.py --path ./data/  && python ./src/GetNodeEdgeNum.py --save ./data/ && python ./src/GetObjectEdgeNum.py --save ./data/
```
**Arguments**:

| **Arguments** | **Detail** |
| --- | --- |
| **path** | Path of the assciation data among different biomoleculars|
| **save** | Save path of the generated data|



## 2，Calculate the protein attribute features
```
python ./src/ACDencoder.py --path ./data/ --save ./data/
```
**Arguments**:

| **Arguments** | **Detail** |
| --- | --- |
| **path** | Path of the protein sequence data|
| **save** | Save path of the generated data|


## 3，Calculate the protein graph features
```
python -m openne --model deepwalk --dataset ./data/AllNodeEdge.csv --num-paths 10 --path-length 80 --window 10

```
**Arguments**:

| **Arguments** | **Detail** |
| --- | --- |
| **dataset** | The graph data. |
| **num-paths** | Number of random walks that starts at each node. |
| **path-length** | Length of random walk started at each node. |
| **window** | Window size of skip-gram model.  |



## 4，Training and prediction under 5-fold cross-validation
```
python ./src/training.py && python ./src/roc_plot.py && python ./src/pr_plot.py
```








# Contributing

All authors were involved in the conceptualization of the proposed method. LWX and SLP conceived and supervised
the project. BYJ and HTZ designed the study and developed the approach. BYJ and HTZ implemented and applied the method to microbial data. BYJ and HTZ analyzed the results. LWX and SLP contributed to the review of the manuscript before submission for publication. All authors read and approved the final manuscript.



# Contacts
If you have any questions or comments, please feel free to email: byj@hnu.edu.cn, slpeng@hnu.edu.cn.

# License

[MIT Richard McRichface.](../LICENSE)
